File: packages/sdk-typescript/adapters/openai.mjs
/**
 * Xurface x OpenAI (function / tool calling) adapter.
 *
 * OpenAI models return `tool_calls`; your app runs each tool by name and hands
 * back a tool message. This adapter inserts Horizon between the model's request
 * and the real tool body: every tool call is recorded, and a call whose risk
 * exceeds the person's appetite is held for their discernment in Xurface Discern
 * before the tool ever runs. A denied or still-pending call comes back as an
 * ordinary tool message, so the model simply sees that it could not proceed.
 *
 * Works with the `openai` SDK and with the Agents SDK - anywhere you dispatch
 * `tool_calls`. The adapter does not call OpenAI itself; it guards the tools.
 *
 * @example
 *   const guard = createOpenAIGuard(xf, { user: xid, agent: "apply-bot", wait: 120000 });
 *   const tools = [{
 *     name: "send_reply", description: "Email a recruiter back",
 *     parameters: { type: "object", properties: { to: {type:"string"}, body:{type:"string"} } },
 *     capability: "reply.send",
 *     run: async ({ to, body }) => sendEmail(to, body),
 *   }];
 *   const reg = guard.register(tools);
 *   // ... send reg.openaiTools in your chat.completions request ...
 *   for (const call of message.tool_calls) toolMessages.push(await reg.dispatch(call));
 */

import { runGuarded } from "../xurface.mjs";

/**
 * @param {import("../xurface.mjs").Xurface} xf
 * @param {{user:string, agent:string, wait?:number}} ctx
 */
export function createOpenAIGuard(xf, ctx) {
  return {
    /**
     * @param {Array<{name:string, description?:string, parameters?:object,
     *   capability?:string, run:(args:any)=>any}>} tools
     */
    register(tools) {
      const byName = new Map(tools.map((t) => [t.name, t]));

      /** The tools array to put in your OpenAI request (schema only). */
      const openaiTools = tools.map((t) => ({
        type: "function",
        function: { name: t.name, description: t.description || "", parameters: t.parameters || { type: "object", properties: {} } },
      }));

      /**
       * Guard + run one tool_call. Returns an OpenAI tool message.
       * @param {{id:string, function:{name:string, arguments:string}}} toolCall
       */
      async function dispatch(toolCall) {
        const name = toolCall.function?.name;
        const tool = byName.get(name);
        let args = {};
        try { args = JSON.parse(toolCall.function?.arguments || "{}"); } catch { /* leave {} */ }
        if (!tool) {
          return toolMessage(toolCall.id, name, { error: `unknown tool ${name}` });
        }
        const out = await runGuarded(xf, {
          user: ctx.user, agent: ctx.agent, capability: tool.capability || name,
          args, wait: ctx.wait,
        }, (finalArgs) => tool.run(finalArgs));

        if (out.ok) return toolMessage(toolCall.id, name, { ok: true, result: out.result, xurface: xurfaceMeta(out.guard) });
        return toolMessage(toolCall.id, name, { ok: false, blocked: true, reason: out.message, xurface: xurfaceMeta(out.guard) });
      }

      return { openaiTools, dispatch, tools: byName };
    },
  };
}

function toolMessage(tool_call_id, name, payload) {
  return { role: "tool", tool_call_id, name, content: JSON.stringify(payload) };
}

function xurfaceMeta(guard) {
  return { state: guard.state, severity: guard.severity, intent: guard.intentId, reasons: guard.reasons };
}

File: packages/sdk-python/xurface/adapters/openai.py
"""Xurface x OpenAI (function / tool calling) adapter - Python.

OpenAI models return `tool_calls`; your app runs each tool by name and returns a
tool message. This adapter inserts Horizon between the model's request and the
real tool body: every call is recorded, and a call whose risk exceeds the
person's appetite is held for their discernment in Xurface Discern before the
tool runs. A held / denied call comes back as an ordinary tool message.

Works with the `openai` SDK and the Agents SDK - anywhere you dispatch
`tool_calls`. The adapter does not call OpenAI; it guards the tools.

Example
-------
    guard = OpenAIGuard(xf, user=xid, agent="apply-bot", wait_ms=120000)
    guard.register("send_reply", send_reply, capability="reply.send",
                   description="Email a recruiter back",
                   parameters={"type": "object", "properties": {"to": {"type": "string"}}})
    # ... pass guard.openai_tools() in your chat.completions request ...
    for call in message.tool_calls:
        tool_messages.append(guard.dispatch(call))
"""

from __future__ import annotations

import json
from typing import Any, Callable, Dict, Optional

from .. import Xurface, run_guarded


class OpenAIGuard:
    def __init__(self, xf: Xurface, user: str, agent: str, wait_ms: int = 0):
        self.xf = xf
        self.ctx = {"user": user, "agent": agent, "wait_ms": wait_ms}
        self._tools: Dict[str, dict] = {}

    def register(self, name: str, func: Callable[..., Any], capability: Optional[str] = None,
                 description: str = "", parameters: Optional[dict] = None) -> "OpenAIGuard":
        self._tools[name] = {"func": func, "capability": capability or name,
                             "description": description,
                             "parameters": parameters or {"type": "object", "properties": {}}}
        return self

    def openai_tools(self) -> list:
        return [{"type": "function",
                 "function": {"name": n, "description": t["description"], "parameters": t["parameters"]}}
                for n, t in self._tools.items()]

    def dispatch(self, tool_call: Any) -> dict:
        """Guard + run one tool_call (dict or SDK object). Returns an OpenAI tool message."""
        fn = _get(tool_call, "function")
        name = _get(fn, "name")
        call_id = _get(tool_call, "id")
        raw_args = _get(fn, "arguments") or "{}"
        try:
            args = json.loads(raw_args) if isinstance(raw_args, str) else dict(raw_args)
        except Exception:
            args = {}
        tool = self._tools.get(name)
        if not tool:
            return _msg(call_id, name, {"error": f"unknown tool {name}"})

        out = run_guarded(self.xf, user=self.ctx["user"], agent=self.ctx["agent"],
                          capability=tool["capability"],
                          exec_fn=lambda final: tool["func"](**final), args=args,
                          wait_ms=self.ctx["wait_ms"])
        g = out["guard"]
        meta = {"state": g.state, "severity": g.severity, "intent": g.intent_id, "reasons": g.reasons}
        if out["ok"]:
            return _msg(call_id, name, {"ok": True, "result": out["result"], "xurface": meta})
        return _msg(call_id, name, {"ok": False, "blocked": True, "reason": out["message"], "xurface": meta})


def _get(obj: Any, key: str) -> Any:
    if isinstance(obj, dict):
        return obj.get(key)
    return getattr(obj, key, None)


def _msg(tool_call_id: str, name: str, payload: dict) -> dict:
    return {"role": "tool", "tool_call_id": tool_call_id, "name": name, "content": json.dumps(payload)}
